Cython/Compiler/ExprNodes.py

```diff
@@ -14246,6 +14246,33 @@ def analyse_as_type(self, env):
         # for compatibility when used as a return_type_node, have this interface too
         return self.analyse_type_annotation(env)[1]
 
+    def _warn_on_unknown_annotation(self, env, annotation):
+        """Method checks for cases when user should be warned that annotation contains unkonwn types."""
+        if annotation.is_name:
+            # Validate annotation in form `var: type`
+            if not env.lookup(annotation.name):
+                warning(annotation.pos,
+                        "Unknown type declaration '%s' in annotation, ignoring" % self.string.value, level=1)
+        elif annotation.is_attribute and annotation.obj.is_name:
+            # Validate annotation in form `var: module.type`
+            if not env.lookup(annotation.obj.name):
+                # `module` is undeclared
+                warning(annotation.pos,
+                        "Unknown type declaration '%s' in annotation, ignoring" % self.string.value, level=1)
+            elif annotation.obj.is_cython_module:
+                # `module` is cython
+                module_scope = annotation.obj.analyse_as_module(env)
+                if module_scope and not module_scope.lookup_type(annotation.attribute):
+                    warning(annotation.pos,
+                            "Unknown type declaration '%s' in annotation, ignoring" % self.string.value, level=1)
```

Comment:
Isn't this rather an error than a warning? And, don't we generate an error for this somewhere in `ParseTreeTransforms` already?

Reply:
> Isn't this rather an error than a warning?

Depends on the overall logic we want to have. Today logic is "Don't error out in pure python mode but rather fallback to python object".

> don't we generate an error for this somewhere in ParseTreeTransforms already?

No, in master following code is compiled without error (in pure python mode):

```
$ cat test.py
import cython

def increment(x: cython.bar):
    return x

print(increment(5))

$ cython test.py
/Users/matus/dev/cython/Cython/Compiler/Main.py:370: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /Users/matus/dev/cython/test/test.py
  tree = Parsing.p_module(s, pxd, full_module_name)
```

Reply:
> Don't error out in pure python mode but rather fallback to python object

However, we do control the `cython.*` namespace. Everything in there should be a valid name or rejected. It doesn't help users to allow arbitrary names (or typos) for `cython.*` names.

Reply:
> However, we do control the cython.* namespace.

Fair enough. I will modify it to error in this case.

Reply:
I have adjusted the PR to raise an error when `cython.*` is not recognized.